cmd/workspace/apps/run_local.go

```diff
@@ -0,0 +1,228 @@
+
+	return config, spec, nil
+}
+
+func setupApp(cmd *cobra.Command, config *apps.Config, spec *apps.AppSpec, customEnv []string, prepareEnvironment bool) (apps.App, []string, error) {
+	ctx := cmd.Context()
+	cfg := cmdctx.ConfigUsed(ctx)
+	app := apps.NewApp(config, spec)
+	env := auth.ProcessEnv(cfg)
+	if cfg.Profile != "" {
+		env = append(env, "DATABRICKS_CONFIG_PROFILE="+cfg.Profile)
+	}
+
+	appEnv, err := spec.LoadEnvVars(ctx, customEnv)
+	if err != nil {
+		return app, nil, err
+	}
+	env = append(env, appEnv...)
+
+	if prepareEnvironment {
+		err := app.PrepareEnvironment()
+		if err != nil {
+			return app, nil, err
+		}
+	}
+
+	return app, env, nil
+}
+
+func startAppProcess(cmd *cobra.Command, config *apps.Config, app apps.App, env []string, debug bool) (*exec.Cmd, error) {
+	ctx := cmd.Context()
+	specCommand, err := app.GetCommand(debug)
+	if err != nil {
+		return nil, err
+	}
+
+	cmdio.LogString(ctx, "Running command: "+strings.Join(specCommand, " "))
+	appCmd := exec.Command(specCommand[0], specCommand[1:]...)
+	appCmd.Stdin = cmd.InOrStdin()
+	appCmd.Stdout = cmd.OutOrStdout()
+	appCmd.Stderr = cmd.ErrOrStderr()
+
+	appEnvs := apps.GetBaseEnvVars(config)
+	for _, envVar := range appEnvs {
+		env = append(env, envVar.String())
+	}
+
+	appCmd.Env = env
+	appCmd.Dir = config.AppPath
+
+	err = appCmd.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	return appCmd, nil
+}
+
+func setupProxy(ctx context.Context, cmd *cobra.Command, config *apps.Config, w *databricks.WorkspaceClient, port int, debug bool) error {
+	proxy, err := appproxy.New(ctx, config.AppURL)
+	if err != nil {
+		return err
+	}
+
+	me, err := w.CurrentUser.Me(ctx)
+	if err != nil {
+		return err
+	}
+
+	for key, value := range apps.GetXHeaders(me) {
+		proxy.InjectHeader(key, value)
+	}
+
+	proxyAddr := fmt.Sprintf("localhost:%d", port)
+	go func() {
+		cmdio.LogString(ctx, "To access your app go to http://"+proxyAddr)
+		err := proxy.ListenAndServe(proxyAddr)
+		if err != nil {
+			cmd.PrintErrln(err)
+		}
+	}()
+
+	if debug {
+		cmdio.LogString(ctx, "To debug your app, attach a debugger to port "+apps.DEBUG_PORT)
+	}
+
+	return nil
+}
+
+// SIGTERM (not supported on Windows) and SIGINT (Ctrl+C, supported cross-platform)
+// are caught to enable graceful shutdown of the app process.
+func handleGracefulShutdown(appCmd *exec.Cmd) error {
+	done := make(chan error, 1)
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
+
+	go func() {
+		done <- appCmd.Wait()
+	}()
+
+	select {
+	case err := <-done:
+		return err
+	case <-sigChan:
+		if err := appCmd.Process.Signal(os.Interrupt); err != nil {
+			return fmt.Errorf("failed to send interrupt signal: %w", err)
+		}
+
+		select {
+		case err := <-done:
+			return err
+		case <-time.After(SHUTDOWN_TIMEOUT):
+			if err := appCmd.Process.Kill(); err != nil {
+				return fmt.Errorf("failed to kill process: %w", err)
+			}
+			return errors.New("process killed after timeout")
+		}
+	}
```

Comment:
For follow-up: some wrappers accept a double SIGINT to force kill the wrapped process. With the current approach, the user has to wait for 15 seconds if the wrapped process ignores SIGINT.